src/js/common/utils/lodash-replacements/set.js

```diff
@@ -0,0 +1,70 @@
+import deconstructPath from './deconstructPath';
+import cloneDeep from './cloneDeep';
+
+/**
+ * Same as `set`, but uses the level param to determine when to clone and give a more helpful error message.
+ *
+ * @param {Array|Obect} object
+ * @param {Array|string} path
+ * @param {*} value
+ * @param {Number} level  How many times we've recursed
+ * @return {Object} A new object with the appropriate value set
+ */
+function baseSet(path, value, object, level = 0) {
+  let arrayPath = Array.isArray(path) ? path : deconstructPath(path);
+
+  if (!arrayPath.length) {
+    // We're at the end of our path; time to assign
+    return value;
+  }
+
+  // Do some one-time prep work
+  let newObj = object;
+  if (level === 0) {
+    // Only clone the whole object once
+    newObj = cloneDeep(object);
```

Comment:
I don't think you need to deep clone the whole object. `set` should end up creating shallow copies of each object in the path (including the root).

Reply:
Note to self: Test the following

```js
const obj = {
  sub1: {
    foo: 'bar'
  },
  sub2: {
    baz: 'asdf'
  }
}

const copy = set(['sub1', 'foo'], 'newVal', obj);

copy.sub2.baz = 'changed on copy';

console.log(obj.sub2.baz);
```

Reply:
Without cloning, the following test (and others) fail:
```js
it('should not modify original object when changing an existing property', () => {
  const o = {
    a: 'a',
    b: { c: 'c' },
    k: { a: { y: 'f' } },
    g: ['h', 'i', 'j']
  };

  // Perhaps using cloneDeep in here is in bad taste?
  const oCopy = _.cloneDeep(o);

  _.set('b.c', 'd', o);
  expect(o).to.eql(oCopy);
});
```
The test output is as follows:
```
  2) lodash replacements set should not modify original object when changing an existing property:

      AssertionError: expected { Object (a, b, ...) } to deeply equal { Object (a, b, ...) }
      + expected - actual

       {
         "a": "a"
         "b": {
      -    "c": "d"
      +    "c": "c"
         }
         "g": [
           "h"
           "i"

      at Context.<anonymous> (test/common/utils/lodash-replacement.unit.spec.js:207:20)
```
Am I missing something?

Reply:
I think you still need to shallow clone at the top level, just not deep clone. Doing a deep clone defeats the purpose of a function/immutable version of `set`.

If I call `set('b.c', ...)` on the object in your test, the object containing `b` and the object containing `c` should not shallowly equal the previous objects at those levels. But `g`, for example, should be shallowly equal in both the new and the old version.